tools/build_node.js

```diff
@@ -6,85 +6,192 @@ const { filter } = require("./lib/dependencies");
 const { rollupWrite } = require("./lib/bundling.js");
 const log = (...args) => console.log(...args);
 
-async function buildNodeIndex(languages) {
+const safeImportName = (s) => {
+  s = s.replace(/-/g, "_");
+  if (/^\d/.test(s)) s = `L_${s}`;
+  return s;
+};
+
+async function buildESMIndex(name, languages) {
+  const header = `import hljs from './core.js';`;
+  const footer = "export default hljs;";
+
+
+  const registration = languages.map((lang) => {
+    const importName = safeImportName(lang.name);
+    return `import ${importName} from './languages/${lang.name}.js';\n` +
+      `hljs.registerLanguage('${lang.name}', ${importName});`;
+  });
+
+  const index = `${header}\n\n${registration.join("\n")}\n\n${footer}`;
+  await fs.writeFile(`${process.env.BUILD_DIR}/es/${name}.js`, index);
+}
+
+async function buildCJSIndex(name, languages) {
   const header = "var hljs = require('./core');";
   const footer = "module.exports = hljs;";
 
   const registration = languages.map((lang) => {
-    let require = `require('./languages/${lang.name}')`;
-    if (lang.loader) {
-      require = require += `.${lang.loader}`;
-    }
+    const require = `require('./languages/${lang.name}')`;
     return `hljs.registerLanguage('${lang.name}', ${require});`;
   });
 
-  // legacy
-  await fs.writeFile(`${process.env.BUILD_DIR}/lib/highlight.js`,
-    "// This file has been deprecated in favor of core.js\n" +
-    "var hljs = require('./core');\n"
-  );
-
   const index = `${header}\n\n${registration.join("\n")}\n\n${footer}`;
-  await fs.writeFile(`${process.env.BUILD_DIR}/lib/index.js`, index);
+  await fs.writeFile(`${process.env.BUILD_DIR}/lib/${name}.js`, index);
 }
 
-async function buildNodeLanguage(language) {
+async function buildNodeLanguage(language, options) {
+  const EMIT = `function emitWarning() {
+    if (!emitWarning.warned) {
+      emitWarning.warned = true;
+      process.emitWarning(
+        'Using file extension in specifier is deprecated, use "highlight.js/lib/languages/%%%%" instead',
+        'DeprecationWarning'
+      );
+    }
+  }
+  emitWarning();`;
+  const CJS_STUB = `${EMIT}
+    module.exports = require('./%%%%.js');`;
+  const ES_STUB = `${EMIT}
+    import lang from './%%%%.js';
+    export default lang;`;
```

Comment:
nit: you could use a more concise form here.
```suggestion
    export { default } from './%%%%.js';`;
```

Reply:
That looks less clear to me.  Possibly just my being unfamiliar with the syntax. Looks like it would export `{ default ` instead of `default`.